src/Shared/Json/JsonSerializerExtensions.cs

```diff
@@ -18,10 +21,47 @@ public static bool IsValid(this JsonTypeInfo jsonTypeInfo, [NotNullWhen(false)]
 
     public static JsonTypeInfo GetReadOnlyTypeInfo(this JsonSerializerOptions options, Type type)
     {
-        options.MakeReadOnly();
+        options.EnsureConfigured();
+
         return options.GetTypeInfo(type);
     }
 
+    public static void EnsureConfigured(this JsonSerializerOptions options)
+    {
+        if (!options.IsReadOnly)
+        {
+            if (!TrimmingAppContextSwitches.EnsureJsonTrimmability)
+            {
+#pragma warning disable IL2026 // Suppressed in Microsoft.AspNetCore.Http.Extensions.WarningSuppressions.xml
+#pragma warning disable IL3050 // Calling members annotated with 'RequiresDynamicCodeAttribute' may break functionality when AOT compiling.
+                InitializeForReflection(options);
+#pragma warning restore IL3050 // Calling members annotated with 'RequiresDynamicCodeAttribute' may break functionality when AOT compiling.
+#pragma warning restore IL2026 // Suppressed in Microsoft.AspNetCore.Http.Extensions.WarningSuppressions.xml
+            }
+
+            options.MakeReadOnly();
+        }
+    }
+
+    [RequiresDynamicCode("JSON serialization and deserialization might require types that cannot be statically analyzed and might need runtime code generation. Ensure Microsoft.AspNetCore.EnsureJsonTrimmability=true.")]
+    [RequiresUnreferencedCode("JSON serialization and deserialization might require types that cannot be statically analyzed. Ensure Microsoft.AspNetCore.EnsureJsonTrimmability=true.")]
+    private static void InitializeForReflection(JsonSerializerOptions options)
+    {
+        // TODO: How can I avoid this lock?
+        lock (options)
```

Comment:
I don't think the lock is avoidable even if the JsonSerializerOptions is the only non-static reference. The call to MakeReadOnly() also needs to go inside the lock just to prevent it from racing with itself.

Ideally, we'd try to avoid races against any caller trying to use the JsonSerializerOptions on another request thread so the serializer never blows up due to a race condition even if the options are being referenced directly. I think we need to find a single point before we start serving requests to lock this down once. Maybe an IHostedService?

Reply:
I've investigate the IHostedService idea and it might work in most cases, however, it will not work on a direct call to RDF (might be an uncommon scenario)

```c#
var app = WebApplication.Create(args);
app.Run(RequestDelegateFactory.Create(() => new { Message = "my delegate" }).RequestDelegate);
app.Run();
```

Reply:
Lets not make this an IHostedService..